File: clients/py/kascov.py
"""kascov.py — a tiny zero-dependency client for the kascov JSON API.

Python 3.9+, stdlib only (urllib). CORS-open API, no keys.

    from kascov import Kascov
    k = Kascov("testnet-10")
    page = k.coins(limit=100)
    coin = k.coin(page["covenants"][0]["covenant_id"])
    for ev in k.stream():           # live events (SSE), blocks forever
        print(ev["kind"], ev["covenant_id"])

Publishing to PyPI is a separate decision — this file is the whole client.
"""
from __future__ import annotations

import json
import urllib.parse
import urllib.request
from typing import Any, Dict, Iterator, Optional

DEFAULT_BASE = "https://kascov.io"


class Kascov:
    def __init__(self, network: str = "mainnet", base: str = DEFAULT_BASE) -> None:
        self.network = network
        self.base = base.rstrip("/")

    def _get(self, path: str) -> Dict[str, Any]:
        req = urllib.request.Request(
            f"{self.base}{path}", headers={"accept": "application/json", "user-agent": "kascov-py"}
        )
        with urllib.request.urlopen(req, timeout=60) as res:
            return json.load(res)

    def live(self) -> Dict[str, Any]:
        """Small fast feed: stats + chain tip + newest ~150 events."""
        return self._get(f"/data/{self.network}-live.json")

    def coins(
        self,
        limit: Optional[int] = None,
        after_daa: Optional[int] = None,
        after_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """One page of coin summaries, newest first. Pass the previous page's
        next_after_daa / next_after_id to walk older coins."""
        q = {k: v for k, v in {"limit": limit, "after_daa": after_daa, "after_id": after_id}.items() if v is not None}
        qs = f"?{urllib.parse.urlencode(q)}" if q else ""
        return self._get(f"/data/{self.network}.json{qs}")

    def coin(self, covenant_id: str) -> Dict[str, Any]:
        """One coin's full story: events, UTXOs (scripts/reveals), holders."""
        return self._get(f"/data/{self.network}/c/{covenant_id}.json")

    def tx(self, txid: str) -> Dict[str, Any]:
        """Which covenant(s) did this transaction move?"""
        return self._get(f"/data/{self.network}/tx/{txid}.json")

    def address(self, addr_or_pubkey: str) -> Dict[str, Any]:
        """Smart coins an address/pubkey funded, received, or controls."""
        return self._get(f"/data/{self.network}/addr/{urllib.parse.quote(addr_or_pubkey)}.json")

    def digest(self) -> Dict[str, Any]:
        """Last-24h digest: births/moves/burns, value born, headliners."""
        return self._get(f"/data/{self.network}/digest.json")

    def galaxy(self) -> Dict[str, Any]:
        """The whole-network app graph (positions + weighted edges)."""
        return self._get(f"/data/{self.network}/galaxy.json")

    def reorgs(self) -> Dict[str, Any]:
        """Recent chain reorgs the indexer rolled back through."""
        return self._get(f"/data/{self.network}/reorgs.json")

    def templates(self) -> Dict[str, Any]:
        """Contract-type analytics (what's running on this network)."""
        return self._get(f"/data/{self.network}/templates.json")

    def activity(self, range: str = "24h") -> Dict[str, Any]:
        """Births/moves/burns per DAA bucket. range: 1h|6h|24h|48h|all"""
        return self._get(f"/data/{self.network}/activity.json?range={range}")

    def stream(self) -> Iterator[Dict[str, Any]]:
        """Live events (SSE) as an iterator. Hints only — refetch on receipt."""
        req = urllib.request.Request(
            f"{self.base}/data/{self.network}/stream",
            headers={"accept": "text/event-stream", "user-agent": "kascov-py"},
        )
        with urllib.request.urlopen(req, timeout=None) as res:
            for raw in res:
                line = raw.decode("utf-8", "replace").strip()
                if line.startswith("data:"):
                    payload = line[5:].strip()
                    if payload:
                        try:
                            yield json.loads(payload)
                        except json.JSONDecodeError:
                            continue

File: clients/js/kascov.mjs
/* kascov.mjs — a tiny zero-dependency client for the kascov JSON API.
   Works in Node 18+ and the browser (native fetch). CORS is open, no keys.

     import { Kascov } from './kascov.mjs';
     const k = new Kascov('testnet-10');
     const { covenants, next_after_daa } = await k.coins({ limit: 100 });
     const coin = await k.coin(covenants[0].covenant_id);
     for await (const ev of k.stream()) console.log(ev.kind, ev.covenant_id);
     for await (const ev of k.stream({ covenant: id })) ...  // one coin only

   The API needs no token. An optional lane token (minted at kascov.io/lane)
   rides along as an X-Kascov-Lane header on every request; it buys extra
   capacity on the holder lane and nothing else — the anonymous tier keeps
   working without it:

     const k = new Kascov('mainnet', 'https://kascov.io', { laneToken: '...' });

   Publishing to npm is a separate decision — this file is the whole client. */

const DEFAULT_BASE = 'https://kascov.io';

export class Kascov {
  constructor(network = 'mainnet', base = DEFAULT_BASE, { laneToken = '' } = {}) {
    this.network = network;
    this.base = base.replace(/\/$/, '');
    this.laneToken = laneToken;
  }

  /* every request goes through here so the lane header can never be forgotten
     on one endpoint and sent on another */
  #headers(accept) {
    const h = { accept };
    if (this.laneToken) h['x-kascov-lane'] = this.laneToken;
    return h;
  }

  async #get(path) {
    const res = await fetch(`${this.base}${path}`, { headers: this.#headers('application/json') });
    if (!res.ok) throw new Error(`kascov: ${path} → HTTP ${res.status}`);
    return res.json();
  }

  /** Small fast feed: stats + chain tip + newest ~150 events. Poll this. */
  live() { return this.#get(`/data/${this.network}-live.json`); }

  /** One page of coin summaries, newest activity first.
      opts: { limit, afterDaa, afterId } — pass the previous page's
      next_after_daa / next_after_id to walk older coins. */
  coins(opts = {}) {
    const q = new URLSearchParams();
    if (opts.limit != null) q.set('limit', opts.limit);
    if (opts.afterDaa != null) q.set('after_daa', opts.afterDaa);
    if (opts.afterId != null) q.set('after_id', opts.afterId);
    const s = q.toString();
    return this.#get(`/data/${this.network}.json${s ? `?${s}` : ''}`);
  }

  /** One coin's full story: events (payloads, moved-with), UTXOs (scripts,
      reveals, budgets), holders. */
  coin(covenantId) { return this.#get(`/data/${this.network}/c/${covenantId}.json`); }

  /** Which covenant(s) did this transaction move? */
  tx(txid) { return this.#get(`/data/${this.network}/tx/${txid}.json`); }

  /** Smart coins an address/pubkey funded, received, or controls. */
  address(addrOrPubkey) { return this.#get(`/data/${this.network}/addr/${encodeURIComponent(addrOrPubkey)}.json`); }

  /** Last-24h digest: births/moves/burns, value born, headliner coins. */
  digest() { return this.#get(`/data/${this.network}/digest.json`); }

  /** The whole-network app graph (positions + weighted edges). */
  galaxy() { return this.#get(`/data/${this.network}/galaxy.json`); }

  /** Recent chain reorgs the indexer rolled back through. */
  reorgs() { return this.#get(`/data/${this.network}/reorgs.json`); }

  /** Contract-type analytics (what's running on this network). */
  templates() { return this.#get(`/data/${this.network}/templates.json`); }

  /** Births/moves/burns per DAA bucket. range: 1h|6h|24h|48h|all */
  activity(range = '24h') { return this.#get(`/data/${this.network}/activity.json?range=${range}`); }

  /** The SSE URL the worker actually serves: /data/{network}/stream, with an
      optional ?covenant=<64 hex> filter narrowing it to one coin. Extracted so
      tests pin the shape against the route registered in main.rs. */
  streamUrl(covenant) {
    const q = covenant ? `?covenant=${encodeURIComponent(covenant)}` : '';
    return `${this.base}/data/${this.network}/stream${q}`;
  }

  /** Live events as an async iterator (SSE). Yields {covenant_id, kind,
      txid, accepting_daa}. Hints only — refetch details on receipt.
      opts.covenant: exactly 64 hex chars follows ONE coin; anything else is
      a 400 from the server, never a silent firehose. */
  async *stream({ covenant, signal } = {}) {
    const res = await fetch(this.streamUrl(covenant), {
      headers: this.#headers('text/event-stream'),
      signal,
    });
    if (!res.ok || !res.body) throw new Error(`kascov: stream → HTTP ${res.status}`);
    const reader = res.body.getReader();
    const dec = new TextDecoder();
    let buf = '';
    for (;;) {
      const { done, value } = await reader.read();
      if (done) return;
      buf += dec.decode(value, { stream: true });
      let i;
      while ((i = buf.indexOf('\n\n')) >= 0) {
        const chunk = buf.slice(0, i);
        buf = buf.slice(i + 2);
        const data = chunk.split('\n').filter((l) => l.startsWith('data:')).map((l) => l.slice(5).trim()).join('');
        if (!data) continue;
        try { yield JSON.parse(data); } catch { /* keepalive/comment */ }
      }
    }
  }
}

export default Kascov;
